File: workspaces-good/crates/calculator/src/capabilities_addition_analyzer.rs
use calculator_shared::contract_calculator_protocol::CalculatorProtocol;
use calculator_shared::taxonomy_expression_vo::ExpressionVO;
use calculator_shared::taxonomy_result_vo::ResultVO;

pub struct AdditionAnalyzer;

impl CalculatorProtocol for AdditionAnalyzer {
    fn evaluate(&self, expr: &ExpressionVO) -> Option<ResultVO> {
        let value = expr.left + expr.right;
        Some(ResultVO::new(expr.left, &expr.op, expr.right, value))
    }
}

File: workspaces-good/packages/addition/src/capabilities_addition_analyzer.ts
import { parseOperand } from "calculator-shared/src/utility_expression_parser";
import { CalculatorProtocol } from "calculator-shared/src/contract_calculator_protocol";
import { ExpressionVO } from "calculator-shared/src/taxonomy_expression_vo";
import {
  ResultVO,
  createResult,
} from "calculator-shared/src/taxonomy_result_vo";

export class AdditionAnalyzer implements CalculatorProtocol {
  evaluate(expr: ExpressionVO): ResultVO {
    const left = parseOperand(String(expr.left)) ?? expr.left;
    const right = parseOperand(String(expr.right)) ?? expr.right;
    const value = left + right;
    return createResult(expr.left, "+", expr.right, value);
  }
}
